test/unit/org/apache/cassandra/io/sstable/format/GlobalTidyConcurrencyTest.java

```diff
@@ -0,0 +1,152 @@
+
+    @BeforeClass
+    public static void setup()
+    {
+        System.setProperty("cassandra.debugrefcount", "true");
+        System.setProperty("org.jboss.byteman.verbose", "true");
+    }
+
+    @Test
+    @BMRule(name = "Count leaks",
+    targetClass="Ref$State",
+    targetMethod="reportLeak",
+    targetLocation="AT EXIT",
+    action="org.apache.cassandra.io.sstable.format.GlobalTidyConcurrencyTest.leakHappened = true;")
+    public void tidyVsGetRaceTest() throws InterruptedException
+    {
+        int NUM_THREADS = 32;
+        Descriptor desc = createDescriptor();
+        AtomicBoolean exit = new AtomicBoolean(false);
+        CopyOnWriteArrayList<Object> errors = new CopyOnWriteArrayList<>();
+
+        class TestThread extends Thread
+        {
+            public TestThread(int idx)
+            {
+                super("test-thread-" + idx);
+            }
+
+            @Override
+            public void run()
+            {
+                try
+                {
+                    while (!exit.get())
+                    {
+                        Ref<SSTableReader.GlobalTidy> ref = SSTableReader.GlobalTidy.get(desc);
+                        LockSupport.parkNanos(ThreadLocalRandom.current().nextInt(1000000));
+                        Ref<SSTableReader.GlobalTidy> currentTidy = SSTableReader.GlobalTidy.lookup.get(desc);
+                        if (!currentTidy.refers(ref.get()))
+                        {
+                            String error = String.format("GlobalTidy to which we keep reference is different than the GlobalTidy associated with this descriptor in SSTableReader.GlobalTidy.lookup; local=%s, lookup=%s, descriptor=%", ref.get(), currentTidy, desc);
+                            ref.release();
+                            throw new AssertionError(error);
+                        }
+                        ref.release();
+                    }
+                }
+                catch (Throwable e)
+                {
+                    logger.error("Stopping test due to error ", e);
+                    errors.add(e);
+                    exit.set(true);
+                }
+            }
+        }
+
+        Thread[] threads = new Thread[NUM_THREADS];
+
+        IntStream.range(0, NUM_THREADS)
+                 .forEach(idx -> threads[idx] = new TestThread(idx));
+
+        for (Thread thread : threads)
+        {
+            thread.start();
+        }
+
+        int NUM_TICKS = 10;
+        for (int tick = 0; tick < NUM_TICKS && !exit.get(); tick++)
+        {
+            Uninterruptibles.sleepUninterruptibly(1, TimeUnit.SECONDS);
+            logger.info("Tick {}...", tick);
+        }
+        exit.set(true);
+
+        for (Thread thread : threads)
+        {
+            thread.join();
+        }
+
+        if (!errors.isEmpty())
+        {
+            logger.error("Unexpected errors in the test:");
+            errors.forEach(error -> logger.error("Error: ", error));
```

Comment:
I don't believe that's true. I added a bogus throw inside the thread:

```ERROR [main] 2022-04-14 14:05:25,230 GlobalTidyConcurrencyTest.java:138 - Error: 
java.lang.RuntimeException: Test error
	at org.apache.cassandra.io.sstable.format.GlobalTidyConcurrencyTest$1TestThread.run(GlobalTidyConcurrencyTest.java:89)
```

the API is:
```
    void error(String var1);
    void error(String var1, Object var2);
    void error(String var1, Object var2, Object var3);
    void error(String var1, Object... var2);
    void error(String var1, Throwable var2);
```

If what you're saying was true, then only the 5th call would end up in properly logging an exception, which is not true, at least not for logback.

Last but not least:
https://github.com/qos-ch/logback/blob/c34645b320d2b31ccaf0de9bb079391904352a28/logback-classic/src/main/java/ch/qos/logback/classic/spi/LoggingEvent.java#L139-L141

i.e. when no throwable is passed it is extracted from the parameters of `Object` type :)

Reply:
now, this `Object` thing is a remnant of how Cassandra 4.1 looks like, where leaks are reported as `Object`, and I'm putting theses `Object` objects into the errors list.

Reply:
Yes, you're correct - sorry for the noise. I think for some reason I thought the `Object` kept was a `String` message rather than the `Throwable`. Never mind.